File: Patterns/Template Pattern/CSharp/Starbuzz/TeaWithHook.cs
using System;

namespace Starbuzz
{
    public class TeaWithHook : CaffeineBeverageWithHook
    {
        private String? GetUserInput()
        {
            Console.Write("레몬을 넣을까요? (y/n) : ");
            return Console.ReadLine();
        }

        public override void Brew() => Console.WriteLine("찻잎를 우려내는 중");

        public override void AddCondiments() => Console.WriteLine("레몬을 추가하는 중");

        public override bool CustomerWantsCondiments() 
        {
            var answer = GetUserInput() ?? "n";

            if (answer.Equals("y")) return true;
            else return false;
        }
    }
}

File: Patterns/Template Pattern/Cpp/Starbuzz/TeaWithHook.cpp
#pragma once

#include "CaffeineBeverageWithHook.cpp"

using namespace std;

class TeaWithHook : public CaffeineBeverageWithHook {
private:
    char getUserInput() {   
        cout << "홍차에 레몬을 넣을까요? (y/n) : ";
                
        char answer;
        cin >> answer;

        return answer;
    }

public:
    void brew() override {
        cout << "찻잎을 우려내는 중\n";
    }

    void addCondiments() override {
        cout << "레몬을 추가하는 중\n";
    }

    bool customerWantsCondiments() override {
        char answer = getUserInput();

        if (answer == 'y') return true;
        return false;
    }
};